lib/modules.nix

```diff
@@ -772,6 +772,17 @@ rec {
       else {};
   };
 
+  /* A simple deep type-checking function that returns a boolean, for use outside
+     of the module system.
+     Note that calling `type.check` is not enough: see `mkOptionType`'s documentation.
+  */
+  typeCheck = type: value: let
+    merged = mergeDefinitions [ ] type [
+      { file = unknownModule; inherit value; }
+    ];
+    eval = builtins.tryEval (builtins.deepSeq merged.mergedValue null);
+  in eval.success;
```

Comment:
The part before `tryEval` and `deepSeq` could be called `typeChecking`: to return the same value, which is lazily checked on demand. Unlike `fix (a: typeCheck t (f a))`, `typeChecking` would allow `f` to use `a` in parts of its return value.
One alternative naming scheme for these two functions would be `strictCheck` / `lazyCheck`.

Reply:
Sounds good, I was thinking of adding something like that for situations where we want an error message rather than `false`. The only question is whether `typeChecking` should take a file argument or use `unknownModule`.